core/src/backend/render.rs

```diff
@@ -341,16 +341,16 @@ fn decode_jpeg(
         jpeg_decoder::PixelFormat::RGB24 => decoded_data,
         jpeg_decoder::PixelFormat::CMYK32 => decoded_data
             .chunks_exact(4)
-            .flat_map(|chunk| {
-                let c = f32::from(chunk[0]);
-                let m = f32::from(chunk[1]);
-                let y = f32::from(chunk[2]);
-                let k = f32::from(chunk[3]);
-
-                let r = ((255.0 - c) * (255.0 - k) / 255.0) as u8;
-                let g = ((255.0 - m) * (255.0 - k) / 255.0) as u8;
-                let b = ((255.0 - y) * (255.0 - k) / 255.0) as u8;
-                [r, g, b]
+            .flat_map(|cmyk| {
+                let c = 255 - u16::from(cmyk[0]);
+                let m = 255 - u16::from(cmyk[1]);
+                let y = 255 - u16::from(cmyk[2]);
+                let k = 255 - u16::from(cmyk[3]);
+
+                let r = (k * c) / 255;
+                let g = (k * m) / 255;
+                let b = (k * y) / 255;
+                [r as u8, g as u8, b as u8]
```

Comment:
I agree with @torokati44, should these be `(k * c + 127) / 255` so that they round properly? Might be worth comparing colors to the Flash Player's output.

Reply:
TL;DR - After some tests, my conclusion is that Flash always rounds down, but `k` should be calculated as `256 - u16::from(cmyk[3])` rather than `255 - u16::from(cmyk[3])`.

I first generated a CMYK JPEG using the following Rust code:
```rs
fn main() {
    let data = [
        254, 0, 0, 3,
        0, 255, 0, 0,
        0, 0, 255, 0,
        255, 255, 255, 0,
    ];
    let encoder = jpeg_encoder::Encoder::new_file("some.jpeg", 100).unwrap();
    encoder.encode(&data, 2, 2, jpeg_encoder::ColorType::Cmyk).unwrap();
}
```
Then I embedded it into a SWF that prints its pixels using `BitmapData`: [cmyk.zip](https://github.com/ruffle-rs/ruffle/files/8505808/cmyk.zip)

Flash traces the following:
```
[object Object]
-16712452
-130818
-66048
-16777216
```

When loaded in Ruffle, `cmyk` appears to be `[254, 0, 0, 4]` for the first (upper left) pixel, though the original data was `[254, 0, 0, 3]` (probably because of lossy compression). From Flash's traces, we learn that Flash decoded that pixel as `-16712452`, which is `0xff00fcfc` as unsigned, which is `rgba(0, 252, 252, 1.0)`. I guess it was calculated as follows:
```
c = 255 - 254 = 1
m = 255 - 0 = 255
y = 255 - 0 = 255
k = 256 - 4 = 252

r = c * k / 255 = 0.9882352941176471 (rounded down to 0)
g = m * k / 255 = 252.0
b = m * k / 255 = 252.0
```
Notice that `k` is subtracted from `256` and not from `255`, because otherwise `g` and `b` would end up as `251`.

Reply:
Hmm, I wonder if this is due to Flash decoding the K with the original 3 value instead of 4 like `jpeg_decoder` (internal rounding.). In which case it may be better to just stick with 255 and truncate the value as we were before (I guess guess adjusting by 1 with 256 wouldn't work for every color value?)